src/irmin-pack/unix/chunked_suffix.ml

```diff
@@ -0,0 +1,189 @@
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Import
+include Chunked_suffix_intf
+
+module Make (Io : Io.S) (Errs : Io_errors.S with module Io = Io) = struct
+  module Io = Io
+  module Errs = Errs
+  module Ao = Append_only_file.Make (Io) (Errs)
+
+  type chunk = { idx : int; suffix_off : int63; ao : Ao.t }
+
+  let suffix_off_to_chunk_poff c off = Int63.Syntax.(off - c.suffix_off)
+
+  type create_error = Io.create_error
+
+  type open_error =
+    [ Io.open_error
+    | `Closed
+    | `Invalid_argument
+    | `Inconsistent_store
+    | `Read_out_of_bounds ]
+
+  (** A simple container for chunks. *)
+  module Inventory : sig
+    type t
+
+    val v : int -> (int -> chunk) -> t
+    val iter_skip_appendable : (chunk -> unit) -> t -> unit
+    val appendable : t -> chunk
+    val find : off:int63 -> t -> chunk
+
+    val open_ :
+      start_idx:int ->
+      chunk_num:int ->
+      open_chunk:
+        (chunk_idx:int -> is_appendable:bool -> (Ao.t, open_error) result) ->
+      (t, [> open_error ]) result
+  end = struct
+    type t = chunk Array.t
+
+    exception OpenInventoryError of open_error
+
+    let v = Array.init
+
+    let iter_skip_appendable f t =
+      Array.sub t 0 (Array.length t - 1) |> Array.iter f
+
+    let appendable t = Array.get t (Array.length t - 1)
+
+    let find ~off t =
+      let find c =
+        let open Int63.Syntax in
+        let end_poff = Ao.end_poff c.ao in
+        let poff = suffix_off_to_chunk_poff c off in
+        Int63.zero <= poff && poff < end_poff
+      in
+      match Array.find_opt find t with
+      | None -> raise (Errors.Pack_error `Read_out_of_bounds)
+      | Some c -> c
+
+    let open_ ~start_idx ~chunk_num ~open_chunk =
+      let off_acc = ref Int63.zero in
+      let create_chunk i =
+        let suffix_off = !off_acc in
+        let chunk_idx = start_idx + i in
+        let is_appendable = i = chunk_num - 1 in
+        let open_result = open_chunk ~chunk_idx ~is_appendable in
+        match open_result with
+        | Error err -> raise (OpenInventoryError err)
+        | Ok ao ->
+            let chunk_len = Ao.end_poff ao in
+            (off_acc := Int63.Syntax.(suffix_off + chunk_len));
+            { idx = chunk_idx; suffix_off; ao }
+      in
+      try Ok (v chunk_num create_chunk)
+      with OpenInventoryError err ->
+        Error (err : open_error :> [> open_error ])
+  end
+
+  type t = { inventory : Inventory.t }
+
+  (* A lightweight wrapper around creating/opening append only files as chunks *)
+  module Ao_chunk = struct
+    let path = Layout.V4.suffix_chunk
+
+    let create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure =
+      let path = path ~root ~chunk_idx in
+      Ao.create_rw ~path ~overwrite ~auto_flush_threshold ~auto_flush_procedure
+
+    let open_rw ~root ~chunk_idx ~auto_flush_threshold ~auto_flush_procedure
+        ~end_poff ~dead_header_size =
```

Comment:
Aha, you're right -- good catch. This code works with the current PR because there is no split. I'm happy to rework this in this PR or let @icristescu take it in the split PR.

Reply:
I just [fixed this](https://github.com/mirage/irmin/compare/43929d1b048439db9c65ce4ab820757b3e4fdfd0..8f2229f96581fc3dcd742869770c868268bd5825) for this PR (and removed the `Ao_chunk` module in the process :tada:!)